backend/cache/redis.go

```
@@ -65,22 +82,75 @@ func (cache *RedisCache) SetWithFields(fields Fields, value interface{}) error {
 	return cache.Set(convertFieldsToKey(fields), value)
 }
 
+// Atomic increment of a value for specified `key`
+func (cache *RedisCache) Increment(key string) error {
+	err := cache.client.Incr(key).Err()
+	if err != nil {
+		log.WithField("key", key).WithError(err).Error("increment: failed to increment key in cache")
+		return err
+	}
+
+	return nil
+}
+
+// Increment value for specified key where key is created by hashing `fields`
+func (cache *RedisCache) IncrementWithFields(fields Fields) error {
+	return cache.Increment(convertFieldsToKey(fields))
+}
+
+// Atomic decrement of a value for specified `key`
+func (cache *RedisCache) Decrement(key string) error {
+	err := cache.client.Decr(key).Err()
+	if err != nil {
+		log.WithField("key", key).WithError(err).Error("decrement: failed to decrement key in cache")
+		return err
+	}
+
+	return nil
+}
+
+// Decrement value for specified key where key is created by hashing `fields`
+func (cache *RedisCache) DecrementWithFields(fields Fields) error {
+	return cache.Decrement(convertFieldsToKey(fields))
+}
+
 // Get value for specified `key`
 func (cache *RedisCache) Get(key string, value interface{}) (bool, error) {
-	bytes, err := cache.client.Get(key).Result()
+	var err error
+	var val int64
+
+	result, err := cache.client.Get(key).Result()
 	if err == redis.Nil {
 		return false, nil
 	} else if err != nil {
-		log.WithField("key", key).WithError(err).Error("Get: failed to get key from cache")
+		log.WithField("key", key).WithError(err).Error("get: failed to get key from cache")
 		return false, err
 	}
 
-	err = msgpack.Unmarshal([]byte(bytes), value)
+	switch v := value.(type) {
+	case *int64:
+		*v, err = strconv.ParseInt(result, 10, 64)
+	case *int32:
+		val, err = strconv.ParseInt(result, 10, 32)
+		*v = int32(val) // We can just cast it here, because on error val will be = 0
```

For me it is not obvious that we can do such things without checking errors first. This time we can because it is an int, and that why I added the comment so it is obvious why we can do that and you don't need to think why we can ignore errors when making an operation on a val. If val was not an int, but a pointer type or something, then performing operations after an error occurred would cause a runtime error (cause by nil pointer).